### PR TITLE
Infrastructure for the agent implementation in rust.

### PR DESCRIPTION
**Pull request by horned-sphere**
The actual definiton of the agents will be done with derive and procedural macros. An example is included of what these macros will looks like. The actual macros will follow in another PR.

Follows on from #498 so review that first.

### CODE REVIEW

**Review by SirCipher — APPROVED** [review]
Very nice :)

File: swim_agent/src/lanes/map/queues/mod.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        I: Iterator<Item = K>,
+    {
+        SyncQueue {
+            id,
+            queue: it.collect(),
+        }
+    }
+
+    pub fn remove(&mut self, key: &K) {
+        //TODO It might be more efficient to do this with a linked list (profile any change).
```
[the remark this answers is not in the document]
Comment (horned-sphere) [reply swim_agent/src/lanes/map/queues/mod.rs:42]: I don't even remember why i did it like that...

File: swim_agent/src/lanes/command/mod.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+mod tests;
+
+/// Model of a command lane. An event if triggered when a command is received (either externally or
+/// internally) but the lane does not maintain any record of its state.
+#[derive(Debug)]
+pub struct CommandLane<T> {
+    id: u64,
+    prev_command: RefCell<Option<T>>,
+    dirty: Cell<bool>,
+    //sync_queue: RefCell<VecDeque<Uuid>>, TODO Is syncing reasonable?
```
[the remark this answers is not in the document]
Comment (horned-sphere) [reply swim_agent/src/lanes/command/mod.rs:42]: I'm not sure syncing with a stateless lane really makes sense (I think we allowed it in the old code). I decided to leave it out for now as it's a bit of a pain with how I did value type uplinks.

File: swim_agent/src/event_handler/mod.rs [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    }
+
+    /// Create a new handler that executes this handler and then performs a side effect.
+    fn followed_by_eff<F, R>(self, eff: F) -> FollowedBy<Self, SideEffect<F>>
+    where
+        Self: Sized,
+        F: FnOnce() -> R,
+    {
+        FollowedBy::new(self, eff.into())
+    }
```
[the remark this answers is not in the document]
Comment (horned-sphere) [reply swim_agent/src/event_handler/mod.rs:92]: I could pull them out. I was trying to reduce the number of things you'd need to import when writing an agent.

File: api/swim_api/src/error/mod.rs [hunk 4]
```diff
@@ -86,15 +86,28 @@ impl From<oneshot::error::RecvError> for AgentRuntimeError {
...
 pub enum AgentInitError {
     #[error("The agent failed to start.")]
     FailedToStart,
+    #[error("Multiple lanes with the same name: {0}")]
+    DuplicateLane(Text),
+    #[error("Error in use code (likely an event handler): {0}")]
+    UserCodeError(Box<dyn std::error::Error>),
+}
+
+//TODO Make this more sophisticated.
```
[the remark this answers is not in the document]
Comment (horned-sphere) [reply api/swim_api/src/error/mod.rs:108]: It's tied in with the thing you noticed about the errors being uninformative in the previous PR. Improving the errors should be a future task.

File: swim_agent/src/agent_model/mod.rs [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+mod tests;
+
+use io::{LaneReader, LaneWriter};
+
+/// Response from a lane after it has written bytes to its outgoing buffer.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WriteResult {
+    // The lane has no data to write (the buffer is still empty).
+    NoData,
+    // The lane has written data to the buffer but has not more data to write.
```
Comment (DobromirM) [inline swim_agent/src/agent_model/mod.rs:54]:
```suggestion
    // The lane has written data to the buffer but has no more data to write.
```

File: swim_agent/src/agent_model/mod.rs [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+    Done,
+    // The lane has written data to the buffer and will have more to write on a subsequent call.
+    DataStillAvailable,
+}
+
+/// At trait which describes the lanes of an agent which can be run as a task attached to an
```
Comment (DobromirM) [inline swim_agent/src/agent_model/mod.rs:60]:
```suggestion
/// A trait which describes the lanes of an agent which can be run as a task attached to an
```

File: swim_agent/src/lanes/command/lifecycle/mod.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+impl<T, Context, Shared, L> CommandLaneLifecycleShared<T, Context, Shared> for L where
+    L: for<'a> CommandLaneHandlersShared<'a, T, Context, Shared>
+{
+}
+
+/// A lifecycle for a command lane with some shared state (shard with other lifecycles in the same agent).
+///
+/// #Type Parameters
+/// * `Context` - The contect for the event handlers (providing access to the agent lanes).
```
Comment (DobromirM) [inline swim_agent/src/lanes/command/lifecycle/mod.rs:69]:
```suggestion
/// * `Context` - The context for the event handlers (providing access to the agent lanes).
```

File: swim_agent/src/lanes/command/mod.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    meta::AgentMetadata,
+};
+
+use super::ProjTransform;
+
+pub mod lifecycle;
+#[cfg(test)]
+mod tests;
+
+/// Model of a command lane. An event if triggered when a command is received (either externally or
```
Comment (DobromirM) [inline swim_agent/src/lanes/command/mod.rs:37]:
```suggestion
/// Model of a command lane. An event is triggered when a command is received (either externally or
```

File: swim_agent/src/lanes/map/lifecycle/mod.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    L: OnUpdateShared<'a, K, V, Context, Shared>
+        + OnRemoveShared<'a, K, V, Context, Shared>
+        + OnClearShared<'a, K, V, Context, Shared>
+{
+}
+
+/// A lifecycle for a map lane with some shared state (shard with other lifecycles in the same agent).
+///
+/// #Type Parameters
+/// * `Context` - The contect for the event handlers (providing access to the agent lanes).
```
Comment (DobromirM) [inline swim_agent/src/lanes/map/lifecycle/mod.rs:87]:
```suggestion
/// * `Context` - The context for the event handlers (providing access to the agent lanes).
```

File: swim_agent/src/lanes/map/mod.rs [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+                result: (),
+            }
+        } else {
+            StepResult::after_done()
+        }
+    }
+}
+
+/// An [`EventHandler`] that will remove an entry from the map.
+
```
Comment (DobromirM) [inline swim_agent/src/lanes/map/mod.rs:326]:
```suggestion
```

File: swim_agent/src/lanes/map/mod.rs [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+                result: (),
+            }
+        } else {
+            StepResult::after_done()
+        }
+    }
+}
+
+/// An [`EventHandler`] that will clear the map.
+
```
Comment (DobromirM) [inline swim_agent/src/lanes/map/mod.rs:363]:
```suggestion
```

File: swim_agent/src/lanes/value/lifecycle/mod.rs [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+impl<'a, L, T, Context, Shared> ValueLaneHandlersShared<'a, T, Context, Shared> for L where
+    L: OnEventShared<'a, T, Context, Shared> + OnSetShared<'a, T, Context, Shared>
+{
+}
+
+/// A lifecycle for a value lane with some shared state (shard with other lifecycles in the same agent).
+///
+/// #Type Parameters
+/// * `Context` - The contect for the event handlers (providing access to the agent lanes).
```
Comment (DobromirM) [inline swim_agent/src/lanes/value/lifecycle/mod.rs:81]:
```suggestion
/// * `Context` - The context for the event handlers (providing access to the agent lanes).
```

File: swim_agent/src/lifecycle/lane_event/mod.rs [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+/// this trait will typically consist of a type level tree (implementations of [`HTree`]) of handlers
+/// for each lane.
+pub trait LaneEvent<'a, Context> {
+    type LaneEventHandler: EventHandler<Context, Completion = ()> + Send + 'a;
+
+    /// Create the handler for a lane, if it exists. It is the responsibility of the lanes to keep track
+    /// of which what events need to be triggered. If the lane does not exist or no event is pending, no
+    /// handler will be returned.
+    /// #Arguments
+    /// * `context` - The content of the agent (allowig access to the lanes).
```
Comment (DobromirM) [inline swim_agent/src/lifecycle/lane_event/mod.rs:45]:
```suggestion
    /// * `context` - The context of the agent (allowing access to the lanes).
```

File: swim_agent/src/lifecycle/lane_event/mod.rs [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+pub trait LaneEventShared<'a, Context, Shared> {
+    type LaneEventHandler: EventHandler<Context, Completion = ()> + Send + 'a;
+
+    /// Create the handler for a lane, if it exists. It is the responsibility of the lanes to keep track
+    /// of which what events need to be triggered. If the lane does not exist or no event is pending, no
+    /// handler will be returned.
+    /// #Arguments
+    /// * `shared` - The shared state.
+    /// * `handler_context` - Utility for constructing event handlers.
+    /// * `context` - The content of the agent (allowig access to the lanes).
```
Comment (DobromirM) [inline swim_agent/src/lifecycle/lane_event/mod.rs:62]:
```suggestion
    /// * `context` - The context of the agent (allowing access to the lanes).
```

**Review by DobromirM — APPROVED** [review]

**Pull request merged**